client/allocrunner/consul_hook_test.go

```diff
@@ -0,0 +1,214 @@
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/nomad/ci"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/consul"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/client/widmgr"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/hashicorp/nomad/nomad/mock"
+	"github.com/hashicorp/nomad/nomad/structs"
+	structsc "github.com/hashicorp/nomad/nomad/structs/config"
+	"github.com/shoenig/test/must"
+)
+
+// statically assert consul hook implements the expected interfaces
+var _ interfaces.RunnerPrerunHook = (*consulHook)(nil)
+
+func testHarness(t *testing.T) *consulHook {
+	logger := testlog.HCLogger(t)
+
+	alloc := mock.Alloc()
+	task := alloc.LookupTask("web")
+	task.Consul = &structs.Consul{
+		Cluster: "default",
+	}
+	task.Identities = []*structs.WorkloadIdentity{
+		{Name: fmt.Sprintf("%s_default", structs.ConsulTaskIdentityNamePrefix)},
+	}
+	task.Services = []*structs.Service{
+		{
+			Provider: structs.ServiceProviderConsul,
+			Identity: &structs.WorkloadIdentity{Name: "consul-service_webservice", Audience: []string{"consul.io"}},
+			Cluster:  "default",
+			Name:     "webservice",
+			TaskName: "web",
+		},
+	}
+
+	identitiesToSign := []*structs.WorkloadIdentity{}
+	identitiesToSign = append(identitiesToSign, task.Identities...)
+	for _, service := range task.Services {
+		identitiesToSign = append(identitiesToSign, service.Identity)
+	}
+
+	// setup mock signer and sign the identities
+	mockSigner := widmgr.NewMockWIDSigner(identitiesToSign)
+	signedIDs, err := mockSigner.SignIdentities(1, []*structs.WorkloadIdentityRequest{
+		{
+			AllocID: alloc.ID,
+			WIHandle: structs.WIHandle{
+				WorkloadIdentifier: task.Name,
+				IdentityName:       task.Identities[0].Name,
+			},
+		},
+		{
+			AllocID: alloc.ID,
+			WIHandle: structs.WIHandle{
+				WorkloadIdentifier: task.Services[0].Name,
+				IdentityName:       task.Services[0].Identity.Name,
+				WorkloadType:       structs.WorkloadTypeService,
+			},
+		},
+	})
+	must.NoError(t, err)
+
+	mockWIDMgr := widmgr.NewMockWIDMgr(signedIDs)
+
+	consulConfigs := map[string]*structsc.ConsulConfig{
+		"default": structsc.DefaultConsulConfig(),
+	}
+
+	hookResources := cstructs.NewAllocHookResources()
+
+	consulHookCfg := consulHookConfig{
+		alloc:                   alloc,
+		allocdir:                nil,
+		widmgr:                  mockWIDMgr,
+		consulConfigs:           consulConfigs,
+		consulClientConstructor: consul.NewMockConsulClient,
+		hookResources:           hookResources,
+		logger:                  logger,
+	}
+	return newConsulHook(consulHookCfg)
+}
+
+func Test_consulHook_prepareConsulTokensForTask(t *testing.T) {
+	ci.Parallel(t)
+
+	hook := testHarness(t)
+	task := hook.alloc.LookupTask("web")
+
+	tests := []struct {
+		name        string
+		task        *structs.Task
+		tokens      map[string]map[string]string
+		wantErr     bool
+		errMsg      string
+		emptyTokens bool
```

Comment:
Because there are a bunch of other identities on this job, maybe we should assert that we got tokens for _specific_ identities and not just that we got some tokens at all? (Same for the test below.)